File: go-backend-interview-prep/02_data_structures/05_strings_and_runes/main.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	// --- Byte length vs rune count ---
	s := "Hello, world"
	fmt.Println("string:", s)
	fmt.Println("len (bytes):", len(s))
	fmt.Println("rune count:", len([]rune(s)))

	// --- Multi-byte characters ---
	fmt.Println("\n--- Multi-byte ---")
	jp := "Go言語" // "Go" + 2 kanji (3 bytes each)
	fmt.Println("string:", jp)
	fmt.Println("len (bytes):", len(jp))       // 8
	fmt.Println("rune count:", len([]rune(jp))) // 4

	// --- Indexing gives bytes, not runes ---
	fmt.Println("\n--- Byte indexing ---")
	fmt.Printf("jp[0] = %c (byte)\n", jp[0]) // 'G'
	fmt.Printf("jp[2] = %d (byte, not a full rune)\n", jp[2])

	// --- Range iterates by rune ---
	fmt.Println("\n--- Range (by rune) ---")
	for i, r := range jp {
		fmt.Printf("  byte_index=%d rune=%c (U+%04X)\n", i, r, r)
	}
	// Notice: byte_index jumps from 2 to 5 to 8 (kanji are 3 bytes)

	// --- Convert to []rune for character manipulation ---
	fmt.Println("\n--- Rune slice manipulation ---")
	runes := []rune(jp)
	runes[2] = 'X' // replace first kanji
	fmt.Println("modified:", string(runes))

	// --- Convert to []byte ---
	fmt.Println("\n--- Byte slice ---")
	b := []byte("hello")
	b[0] = 'H'
	fmt.Println("modified bytes:", string(b))

	// --- strings.Builder for efficient concatenation ---
	fmt.Println("\n--- strings.Builder ---")
	var sb strings.Builder
	for i := 0; i < 5; i++ {
		fmt.Fprintf(&sb, "item%d ", i)
	}
	fmt.Println("built:", sb.String())

	// --- Useful string functions ---
	fmt.Println("\n--- String functions ---")
	msg := "  Go is Great  "
	fmt.Println("TrimSpace:", "'"+strings.TrimSpace(msg)+"'")
	fmt.Println("Contains:", strings.Contains(msg, "Great"))
	fmt.Println("Split:", strings.Split("a,b,c", ","))
	fmt.Println("Join:", strings.Join([]string{"x", "y", "z"}, "-"))

	// --- Rune literal vs string literal ---
	fmt.Println("\n--- Rune vs string literal ---")
	var r rune = 'A'
	var str string = "A"
	fmt.Printf("rune 'A': type=%T value=%d\n", r, r)
	fmt.Printf("string \"A\": type=%T len=%d\n", str, len(str))
}

File: go-backend-interview-prep/02_data_structures/05_strings_and_runes/main.py
# Python equivalent of Strings and Runes (compare with main.go)
# Python str is Unicode (not bytes). len() returns characters.

def main():
    # --- Length = character count (unlike Go!) ---
    s = "Hello, world"
    print("string:", s)
    print("len (characters):", len(s))  # Go: len() returns bytes
    print("len (bytes):", len(s.encode("utf-8")))

    # --- Multi-byte characters ---
    print("\n--- Multi-byte ---")
    jp = "Go\u8a00\u8a9e"  # "Go言語"
    print("string:", jp)
    print("len (characters):", len(jp))              # 4
    print("len (bytes):", len(jp.encode("utf-8")))   # 8

    # --- Indexing gives a character (unlike Go which gives a byte) ---
    print("\n--- Character indexing ---")
    print(f"jp[0] = {jp[0]}")   # 'G'
    print(f"jp[2] = {jp[2]}")   # '言' (full character, unlike Go)

    # --- Iterate by character (default in Python) ---
    print("\n--- Iterate (by character) ---")
    for i, ch in enumerate(jp):
        print(f"  index={i} char={ch} (U+{ord(ch):04X})")

    # --- Strings are immutable (same as Go) ---
    # s[0] = 'h'  # TypeError

    # --- Modify: convert to list of chars ---
    print("\n--- Modify via list ---")
    chars = list(jp)
    chars[2] = 'X'
    print("modified:", "".join(chars))

    # --- bytes vs str ---
    print("\n--- bytes vs str ---")
    b = b"hello"      # bytes literal
    print("bytes:", b)
    print("decoded:", b.decode("utf-8"))

    # --- Building strings efficiently ---
    print("\n--- String building ---")
    parts = []
    for i in range(5):
        parts.append(f"item{i}")
    print("built:", " ".join(parts))

    # --- Useful string methods ---
    print("\n--- String methods ---")
    msg = "  Python is Great  "
    print("strip:", repr(msg.strip()))
    print("contains:", "Great" in msg)
    print("split:", "a,b,c".split(","))
    print("join:", "-".join(["x", "y", "z"]))


if __name__ == "__main__":
    main()
